docs/conditional-steps.md

```diff
@@ -0,0 +1,61 @@
+# Conditional Steps
+
+Conditional steps allow the definition of steps that only run if a condition is
+met. For example, an orb could define a command that runs a set of steps if the
+orb's user invokes it with `myorb/foo: { dostuff: true }`, but not
+`myorb/foo: { dostuff: false }`.
+
+Note that the condition is checked before a workflow is actually run. This
+means, for example, that a user can't use a condition to check an environment
+variable.
+
+Conditional steps can be located anywhere a regular step could. For example, an
+orb author could define conditional steps in the `steps` key of a Job or a
+Command.
+
+A conditional step consists of a step with the key `when` or `unless`. The
+subkey `steps` defines the steps to run if the condition under subkey
+`condition` is met. Currently the `condition` is a single value that evaluates
+to `true` or `false`.
+
+### Conditional steps in an orb
+
+```
+# inside orb.yml for orb `myorb`
+
+jobs:
+  myjob:
+    parameters:
+      preinstall-foo:
+        type: boolean
+        default: false
+      foo-download-url:
+        type: string
+        default: ""
+    executor: some-executor
+    steps:
+      - checkout
+      - when:
+        condition: << parameters.preinstall-foo >>
+        steps:
+          - install-foo:
+            download-url: << parameters.foo-download-url >>
+      - unless:
+        condition: << parameters.preinstall-foo >>
+        steps:
+          - compile-foo
+      - verify-foo-version
+      - build
+      - test
+```
+
+```
+# inside user's build.yml
+workflows:
+  workflow:
+    jobs:
+      - myorb/myjob:
+          preinstall-foo: true
+          foo-download-url: https://example.com/foo.sh
+```
```

Comment:
I wasn't able to get this working with the files in `~/tmp/build.yml`, `~/tmp/orbs/myorb/orb.yml` and `orb-compiler-alpha compile --path ~/tmp`.  However, was able to compile this single `build.yml`:

```
version: 2
workflows:
  workflow:
    jobs:
      - myjob:
          preinstall-foo: false
      - myjob:
          preinstall-foo: true

jobs:
  myjob:
    parameters:
      preinstall-foo:
        type: boolean
        default: false
    machine: true
    steps:
      - run: echo "preinstall is << parameters.preinstall-foo >>"
      - when:
          condition: << parameters.preinstall-foo >>
          steps:
            - run: echo "preinstall"
      - unless:
          condition: << parameters.preinstall-foo >>
          steps:
            - run: echo "don't preinstall"
```

That compiled into this:

```
version: 2
jobs:
  myjob-1:
    machine: true
    steps:
    - run:
        command: echo "preinstall is false"
    - run:
        command: echo "don't preinstall"
  myjob-2:
    machine: true
    steps:
    - run:
        command: echo "preinstall is true"
    - run:
        command: echo "preinstall"
workflows:
  workflow:
    jobs:
    - myjob-1
    - myjob-2
```

I think it's better to include currently working config over planned config.